software/contrib/turing_machine.py

```diff
@@ -0,0 +1,254 @@
+        the `max_output_voltage`. Represented by a float in the range [0, `max_output_voltage`]"""
+        return self.scale_getter()
+
+    @scale.setter
+    def scale(self, scale):
+        """Set the scale factor as a float in the range [0, `max_output_voltage`]"""
+        if scale < 0 or scale > self.max_output_voltage:
+            raise ValueError(f"Scale of {scale} is outside the expected range of [0,{self.max_output_voltage}]")
+        self._scale = scale
+
+    @property
+    def length(self):
+        """Returns the length of the current sequence as an integer in the range of [2, `bit_count`]"""
+        return self.length_getter()
+
+    @length.setter
+    def length(self, length):
+        """Sets the length of the current sequence as an integer in the range of [2, `bit_count`]"""
+        if length < 2 or length > self.bit_count:
+            raise ValueError(f"Length of {length} is outside the expected range of [2,{self.bit_count}]")
+        self._length = length
+
+    @property
+    def write(self):
+        """Returns the current value of the 'write switch'. When true the least significant bit will be cleared during
+        rotation, regardless of the `flip_probability`. This allows for real-time user manipulation of the sequence."""
+        return self.write_getter()
+
+    @write.setter
+    def write(self, value: bool):
+        """Set the state of the 'write switch'. `True` means that the least significant bit will be cleared."""
+        self._write = value
+
+
+class EuroPiTuringMachine(EuroPiScript):
+    def __init__(self, bit_count=DEFAULT_BIT_COUNT, max_output_voltage=MAX_OUTPUT_VOLTAGE):
+        self.tm = TuringMachine(bit_count, max_output_voltage)
+        self.tm.flip_probability_getter = self.flip_probability
+        self.tm.scale_getter = self.scale
+        self.tm.length_getter = self.length
+        self.tm.write_getter = self.write
+        self.tm.step_handler = self.step_handler
+        self.k2_scale_mode = True  # scale mode or length mode
+
+        @din.handler
+        def clock():
+            self.tm.step()
+
+        @b2.handler_falling
+        def mode_toggle():
+            if self.k2_scale_mode:
+                self.tm.scale = self.scale()
+            else:
+                self.tm.length = self.length()
+            self.k2_scale_mode = not self.k2_scale_mode
+
+    def step_handler(self):
+        cv1.value(self.tm.get_bit(CV1_PULSE_BIT))
+        cv2.value(self.tm.get_bit(CV2_PULSE_BIT))
+        cv3.value(self.tm.get_bit(CV3_PULSE_BIT))
+        cv4.value(self.tm.get_bit_and(CV1_PULSE_BIT, CV2_PULSE_BIT))
+        cv5.value(self.tm.get_bit_and(CV2_PULSE_BIT, CV3_PULSE_BIT))
+        cv6.voltage(self.tm.get_voltage())
+
+    def flip_probability(self):
+        return clamp(int((round(1 - k1.percent() - ain.percent(), 2)) * 100), 0, 100)
+
+    def scale(self):
+        if self.k2_scale_mode:
+            return k2.percent() * self.tm.max_output_voltage
```

Comment:
(I just had a long explanation of this suggestion typed up, but I clicked on a link and GitHub didn't save the draft :sob:)

I noticed that when changing which parameter receives k2 value, the attribute takes the value of the current knob position, which may cause it to jump to an undesired value. You can avoid this by storing the previous knob value and only set the attribute value when k2 value changes.

When setting the attribute value, only set the value if the knob value has changed from the previous loop iteration. When prev value is None, this is the first loop iteration and we should not set the attribute value, but should capture the value according to the current knob position.

https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/polyrhythmic_sequencer.py#L259-L263

When switching attributes with b2, set the previous knob value to None since the data type or value range may be different across attributes.

https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/polyrhythmic_sequencer.py#L156-L159

Reply:
I took this a step further and added a couple new classes to handle this situation. They are in their own PR (#155) and I have updated this PR to use them. Let me know what you think.